File: dev/requirements/behavior/cases/interruption-resumes.6.6.case.swift
import Foundation
import SetlistCore

/// 6.6 — An audio-session interruption finalises the recording so far and
/// resumes into the same show without losing captured audio.
let interruptionResumes_6_6 = RequirementCase(id: "6.6") {
    let session = FakeRecordingSession()
    let coordinator = CaptureCoordinator(session: session, container: FileManager.default.temporaryDirectory)
    let showID = UUID()

    try coordinator.startShow(showID)
    session.record(120)
    let beforeInterruption = coordinator.now

    // A call arrives mid-set.
    try coordinator.interruptionBegan()
    try expect(!coordinator.isRecording, "the microphone is gone")
    session.gap(30)
    try coordinator.interruptionEnded()
    try expect(coordinator.isRecording, "the show resumed")

    session.record(60)
    let afterInterruption = coordinator.now
    let parts = try coordinator.endShow()

    try expectEqual(parts.count, 2, "the show kept both stretches")
    try expectEqual(session.startedURLs, parts, "every stretch was written somewhere")
    try expect(parts[0] != parts[1], "the second stretch did not overwrite the first")
    try expect(afterInterruption > beforeInterruption, "the clock did not rewind across the gap")
    try expectEqual(afterInterruption, CaptureTime(180), "only recorded seconds count")
}

File: dev/requirements/behavior/cases/interruption-resumes.6.6.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/fake_world.dart';
import '../../shared/requirement_case.dart';

/// 6.6 — An audio-session interruption finalises the recording so far and
/// resumes into the same show without losing captured audio.
final interruptionResumes_6_6 = RequirementCase('6.6', (tester) async {
  final session = FakeRecordingSession();
  final capture = CaptureCoordinator(session: session, containerPath: '/container');

  await capture.startShow('show-1');
  session.recordSeconds(120);

  await capture.interruptionBegan();
  expectFalse(capture.isRecording, 'the part so far is closed rather than abandoned');

  // Nothing is being written, so nothing is being recorded: the wall clock
  // moves, the capture clock does not.
  session.recordSeconds(45);
  expectThat(capture.now.seconds, 120, 'the clock does not advance across the gap');

  await capture.interruptionEnded();
  expectTrue(capture.isRecording, 'the show continues');

  session.recordSeconds(30);
  expectThat(capture.now.seconds, 150, 'and the clock never rewinds');

  final parts = await capture.endShow();
  expectThat(parts.length, 2, 'the show is two stretches of audio, both kept');
});
